### PR TITLE
Disallow untargeted introspection queries

### PR DESCRIPTION
**Pull request by teskje**
See commits for details.

### Motivation

Which of the following best describes the motivation behind this PR?

  * This PR adds a known-desirable feature.

    Part of #12520.

### Tips for reviewer

The first commit splits a new `SourceConnector::Log` variant from `SourceConnector::Local`, to make it possible to identify log sources in the dependency analysis (second commit). This is certainly not the only way to implement this, we could also add a flag to `SourceConnector::Local` or track log source IDs somewhere else in the catalog. Let me know if you think another solution would be better!

The error messages for `UntargetetLogRead` are supposed to be as clear as possible, since we don't plan to add additional documentation for this feature. Suggestions on how to improve these are welcome.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Renames the `mz_sources.connector_type` value of log sources from "local" to "log".

### CODE REVIEW

File: src/coord/src/coord.rs
Comment (teskje) [inline src/coord/src/coord.rs:3091]: There is an edge case where this check can fail do prevent untargeted log peeks. Assume we have a cluster with a single replica and two clients running statements against this cluster.

1. Client A: `SELECT * FROM mz_materializations AS OF <now + 10s>`.
2. Client B: `CREATE CLUSTER REPLICA this_cluster.replica2 ...`.
3. Client A's `SELECT` returns.

Because Client A's `SELECT` was made without `cluster_replica` being set, `replica2` was hydrated with the `Peek` corresponding to that `SELECT`. Which means when the `SELECT` eventually returns, either of the two replicas could have produced the response.

We have two ways to handle this:

1. Accept that we don't handle this edge-case correctly, as it is unlikely to occur anyway.
2. Always set `target_replica` here, if `num_replicas == 1` and we query introspection sources, e.g.:

    ```rust
    let log_names = source_ids
        .iter()
        .flat_map(|id| self.catalog.log_dependencies(*id))
        .map(|id| self.catalog.get_entry(&id).name().item.clone())
        .collect::<Vec<_>>();
    if !log_names.is_empty() {
        if num_replicas == 1 {
            target_replica = compute_instance.replicas_by_id.keys().next().unwrap();
        } else if target_replica.is_none() {
            return Err(CoordError::UntargetedLogRead { log_names });
        }
    }
    ```

I'm not sure if fixing our handling of the above marginal edge-case warrants this additional complexity. Any opinions?

File: src/storage/src/storage_state.rs
Comment (benesch) [inline src/storage/src/storage_state.rs:120]: This probably should just panic if it sees a SourceConnector::Log, as you can’t create a log source manually. We don’t presently ever send a CreateSources command with a log source, so it works out ok now, but given that we now have the language to talk about log and table sources separately, might as well make the distinction in behavior explicit!

**Review by benesch — COMMENTED** [review]
This looks stellar, but leaving the final review to Sean because I’m on mobile. Splitting SourceConnector::Log from SourceConnector::Local seems like a big win for clarity, so thumbs up from me, but if for some reason we didn’t want to do the split, you can determine whether something’s a log source by checking whether its ID is in BUILTIN_LOGS.

File: src/coord/src/coord.rs
Comment (teskje) [reply src/coord/src/coord.rs:3091]: If a replica is dropped that still has targeted peeks waiting, nothing happens immediately. The peek is left in the compute controller's list of outstanding peeks and the user's SQL session hangs on the `SELECT`.

Eventually, we should see a `CancelPeek` when the user aborts the `SELECT` or closes their session (the latter isn't the case currently: https://github.com/MaterializeInc/materialize/issues/12546). The `CancelPeek` removes the outstanding peek from the compute controller's state. Then during the next history reduction, the peek is also removed from the command history.

**Review by antiguru — APPROVED** [review]
LGTM pending Nikhil's and Sean's feedback

File: src/storage/src/storage_state.rs
Comment (teskje) [reply src/storage/src/storage_state.rs:120]: We actually do send `CreateSources` commands for log sources. It happens during bootstrapping the coordinator, where it loads all sources from the catalog, including the built-in ones, and announces them to storage: https://github.com/MaterializeInc/materialize/blob/main/src/coord/src/coord.rs#L558-L582. So if I change that code to panic, materialized isn't able to start up anymore.

I assume we could adjust the coordinator to not send these `CreateSources` commands in the first place, but I'd rather wait for @petrosagg's separation PR to land, since it touches that code anyway.

Comment (benesch) [reply src/storage/src/storage_state.rs:120]: Oh, sigh. I guess we’ve just been incorrectly creating tables for all these logging sources? Can you file a follow up issue and assign to me?

**Review by sploiselle — APPROVED** [review]

**Pull request merged**
